src/class/dfu/dfu_rt_device.h

```diff
@@ -29,36 +29,52 @@
 
 #include "common/tusb_common.h"
 #include "device/usbd.h"
+#include "dfu.h"
 
 #ifdef __cplusplus
  extern "C" {
 #endif
 
-
 //--------------------------------------------------------------------+
-// Common Definitions
+// Application Callback API (weak is optional)
 //--------------------------------------------------------------------+
+// Allow the application to update the status as required.
+// Is set to DFU_STATUS_OK during internal initialization and USB reset
+// Value is not checked to allow for custom statuses to be used
+void tud_dfu_runtime_set_status(dfu_mode_device_status_t status);
 
-// DFU Protocol
-typedef enum
-{
-  DFU_PROTOCOL_RT  = 1,
-  DFU_PROTOCOL_DFU = 2,
-} dfu_protocol_type_t;
+// Invoked when a DFU_DETACH request is received and bitWillDetach is set
+void tud_dfu_runtime_reboot_to_dfu_cb();
 
-// DFU Descriptor Type
-typedef enum
-{
-  DFU_DESC_FUNCTIONAL = 0x21,
-} dfu_descriptor_type_t;
+// Invoked when a DFU_DETACH request is received and bitWillDetach is not set
+// This should start a timer for wTimeout ms
+// When the timer has elapsed, the app must call tud_dfu_runtime_detach_timer_elapsed
+// If a USB reset is called while the timer is running, the class will call
+// tud_dfu_runtime_reboot_to_dfu_cb.
+// NOTE: This callback should return immediately, and not implement the delay
+// internally, as this can will hold up the USB stack
+TU_ATTR_WEAK void tud_dfu_runtime_detach_start_timer_cb(uint16_t wTimeout);
 
+// Invoke when the dfu runtime detach timer has elapsed
+void tud_dfu_runtime_detach_timer_elapsed();
 
-//--------------------------------------------------------------------+
-// Application Callback API (weak is optional)
-//--------------------------------------------------------------------+
+// Invoked when a nonstandard request is received
+// Use may be vendor specific.
+// Return false to stall
+TU_ATTR_WEAK bool tud_dfu_runtime_req_nonstandard_cb(uint8_t rhport, uint8_t stage, tusb_control_request_t const * request);
+
+// Invoked during initialization of the dfu driver to set attributes
+// Return byte set with bitmasks:
+//   DFU_FUNC_ATTR_CAN_DOWNLOAD_BITMASK
+//   DFU_FUNC_ATTR_CAN_UPLOAD_BITMASK
+//   DFU_FUNC_ATTR_MANIFESTATION_TOLERANT_BITMASK
+//   DFU_FUNC_ATTR_WILL_DETACH_BITMASK
+// Note: This should match the USB descriptor
+uint8_t tud_dfu_runtime_init_attrs_cb();
 
-// Invoked when received new data
-TU_ATTR_WEAK void tud_dfu_runtime_reboot_to_dfu_cb(void);
+// Invoked during a DFU_GETSTATUS request to get for the string index
+// to the status description string table.
+TU_ATTR_WEAK uint8_t tud_dfu_runtime_get_status_desc_table_index_cb();
```

Comment:
again, I think runtime only support 2 request: 
1.GET_STATUS which return APP_IDLE and
2.DETACH which reboot immediately.